### ISSUE TITLE
Bot pour query

### ISSUE THREAD

**Issue by iv-stpn** [noftaly/MonkaBot#8]

**Comment by noftaly** [comment]
@iv-stpn Tu peux détailler stp ?

**Comment by iv-stpn** [comment]
@Kethash 

Pour la requête POST il faut mettre les arguments dans le body

Pour la commande en elle-même tu peux mettre les arguments à la suite comme ça: -arg1 valeur -arg2 valeur, etc.

D'ailleurs si tu peux il faudrait aussi faire un moyen de gérer les headers, tu sais commment t'y prendre ?

Sinon, n'hésite pas à demander de l'aide ou des clarifications :)
De préférence directement sur l'issue d'ailleurs

**Comment by noftaly** [comment]
Honnêtement je ne vois pas tellement l'intérêt de cette commande, ca sert a quoi de faire des requêtes http via discord ?
C'est pas du tout pratique, t'as plus vite fait de le faire via ton navigateur directement.
Et si c'est une requête post avec des headers spécial, jsp si discord, et qui plus est le serveur Ef'réussite, est l'endroit le plus adapté pour ca non ?

**Comment by iv-stpn** [comment]
C'est intéressant pour faire des démos lives, pour pull des documents sans avoir à les télécharger puis les drag and drop, et c'est tout aussi pertinent qu'un moteur pour exécuter du code, juste pour le web

Ce n'est pas trivial de simuler une requête POST juste avec ton navigateur sinon. Bien sûr tu peux utiliser Postman, Burp Suite, etc., mais les débutants ne sont pas familiers avec ça forcément et si tu ne l'as pas installé/ouvert déjà c'est plus rapide de faire la commande sur Discord

**Comment by noftaly** [comment]
> pour pull des documents sans avoir à les télécharger puis les drag and drop

Si les documents viennent de moodle, le bot n'aura pas la perm de les voir car il faut se connecter. Idem s'ils viennent du GDrive horizon. Honnêtement je vois rarement un cas ou tu vas avoir besoin de télécharger un doc et ce sera plus rapide de passer par discord... Mais why not

> Ce n'est pas trivial de simuler une requête POST juste avec ton navigateur sinon. Bien sûr tu peux utiliser Postman, Burp Suite, etc., mais les débutants ne sont pas familiers avec ça forcément et si tu ne l'as pas installé/ouvert déjà c'est plus rapide de faire la commande sur Discord

Oui je sais, mais déjà pourquoi tu aurais envie de passer une requête POST, et pourquoi un débutant voudrai passer une requête POST ? Genre tu as quoi comme exemple ? 😅 

Après en soit une commande de plus ou de moins on est pas a ca prêt, et si tu t'en sert même juste une fois ce sera rentabiliser, mais pour moi ca reste le genre de truc que t'utilisera jamais car c'est pas adapté 😄

**Comment by iv-stpn** [comment]
Tu mélanges les deux commandes, le but de cette commande n'est pas de lier quoi que ce soit à Moodle, c'est le but de l'autre commande

Commande POST ou quoi que ce soit d'autre en HTTP, tu peux aussi imaginer des GET avec des headers, c'est dur de les simuler depuis navigateur

Autant avoir plus de commandes que nécessaire que pas assez je pense aussi, mais je pense qu'on trouvera des utilités pour durant les cours de web

Comme je te dis, c'est aussi utile que la commande pour exécuter du code

**Comment by noftaly** [comment]
> Commande POST ou quoi que ce soit d'autre en HTTP, tu peux aussi imaginer des GET avec des headers, c'est dur de les simuler depuis navigateur

Mais encore une fois, donne moi un exemple concret de l'utilité que tu aurais de faire une requête GET/POST avec headers via discord !

> Comme je te dis, c'est aussi utile que la commande pour exécuter du code

Pas vraiment, car ca ca peut être utilisé pour débugger le code de qqn ou leur expliquer qqchose. Faire une requête POST avec headers qui va te retourner un fichier json, je vois mal un cas d'utilisation. Mais si t'as des exemples, vas-y !

**Comment by noftaly** [comment]
Quand j'y pense, tu veux gérer les différentes méthodes (get/post), gérer le body, gérer les headers... C'est juste en train de faire une copie conforme de `curl` 😄
Je répète, pour moi tu n'es clairement pas en train de simplifier l'action, parce que je ne vois pas l'intérêt de lancer curl via discord plutôt que le terminal...

**Comment by iv-stpn** [comment]
Yes c'est ça, c'est quasiment l'idée de faire une version Discord de curl, on peut même utiliser curl dans le bot si c'est plus simple
Les utilités seraient:
1. Permettre aux gens qui n'ont pas curl sur leur machine Windows (a.k.a. 90% des gens qui ont Windows) de tester des requêtes
2. Permettre de pull directement des fichiers plutôt que du texte sur CLI ou un fichier en local qu'il faudra drag and drop
3. Permettre de customiser des routines qui pourraient être liées à des sous-commandes et d'avoir des "recettes" de requêtes prédéfinies avec certains headers par exemple
4. Permettre de lier ces routines à d'autres commandes (moduler un peu les requêtes d'API qui pourraient être faites dans d'autres commandes)
5. Permettre de lier ces routines à des réactions sur des messages, ce genre de chose

L'utilité d'avoir un système pour faire des queries depuis DIscord sera particulièrement utile quand le site tournera, pour pull des fichiers, pour moi c'est l'utilité principale; une autre utilité serait pour que les Profs et étudiants ait un playground directement sur le serveur pour faire des démos/tests

Après c'est sûr qu'on est pas au summum de l'utilité, mais par exemple si tu travailles sur une commande qui doit récupérer des infos d'un API (comme des horaires de train/bus wink wink), tu pourrais passer par une méthode qui fait une surcouche Discord de curl

Tu pourras me dire qu'on pourra à chaque fois faire au cas par cas, parce qu'il existe déjà des façons de faire des queries en Node de façon pas trop difficile, mais autant créer une fonction qui a déjà des routines en fonction de certains cas et qui a une syntaxe plus beginner-friendly que curl

Après c'est en priorité basse quoi qu'il arrive, si tu veux on peut la mettre en attente en attendant que l'API du site soit là

**Comment by noftaly** [comment]
> Permettre aux gens qui n'ont pas curl sur leur machine Windows (a.k.a. 90% des gens qui ont Windows) de tester des requêtes

Le plus simple serait quand même pour eux qu'ils l'installent car ils en auront besoin un jour ou l'autre. curl est quand même un must-have, surtout qu'il est aussi utilisé par beaucoup d'autres CLI (comme git)

> Permettre de pull directement des fichiers plutôt que du texte sur CLI ou un fichier en local qu'il faudra drag and drop

Je ne vois pas l'intérêt, autant mettre l'URL directement 😅 et tu cliques dessus et ca te télécharges le fichier...

> Permettre de customiser des routines qui pourraient être liées à des sous-commandes et d'avoir des "recettes" de requêtes prédéfinies avec certains headers par exemple
Permettre de lier ces routines à d'autres commandes (moduler un peu les requêtes d'API qui pourraient être faites dans d'autres commandes)
Permettre de lier ces routines à des réactions sur des messages, ce genre de chose

Je ne comprends pas... les commandes ne vont pas exécuter d'autres commandes sur discord pour faire une requête 🤔 si une commande doit faire une requête, elle est faite directement en typescript dans le code, avec axios

> utile quand le site tournera, pour pull des fichiers

Autant avoir directement une commande pour interagir avec le site, ce sera beaucoup plus simple, et beaucoup plus beginner friendly pour le coup

> pour que les Profs et étudiants ait un playground directement sur le serveur pour faire des démos/tests

Dans ces cas la on revient à https://github.com/EFREI-Horizon/MonkaBot/issues/8#issuecomment-878567696, autant utiliser le terminal directement c'est beaucoup plus adapté et beaucoup plus simple

> Si tu travailles sur une commande qui doit récupérer des infos d'un API tu pourrais passer par une méthode qui fait une surcouche Discord de curl

Non, tu passes par axios, pas par une autre commande discord:
1. Ca évite des requêtes inutiles discord ("requête à discord pour poster le message de la commande curl, requête de la commande curl à l'URL demandé, requête de la commande curl à discord pour poster la réponse" vs "requête de la commande curl à l'URL demandé, requête de la commande curl à discord pour poster la réponse")
2. Ca va envoyer des messages en plus dans le chat (le message du bot pour exécuter la commande curl)
3. Le bot ne reçoit pas les messages des bots, donc il ne peut pas auto-exécuter la commande curl

> Tu pourras me dire qu'on pourra à chaque fois faire au cas par cas, parce qu'il existe déjà des façons de faire des queries en Node de façon pas trop difficile, mais autant créer une fonction qui a déjà des routines en fonction de certains cas et qui a une syntaxe plus beginner-friendly que curl

Je me demandais si j'avais bien compris car ca me parait absurde de se dire "j'ai besoin de faire une query en Node.js, donc je vais faire une commande discord qui fait une query, et quand j'ai besoin de faire une query en Node.js, je poste un message sur discord pour utiliser la commande discord"; mais j'ai l'impression que c'est bien ce que tu proposes. Je rappelle que faire une query en Node.js c'est littéralement 1 ligne 😅 
```ts
const response = await axios.get('https://example.com');
```


Désolé mais je ne vois pas une seule utilité valide pour le moment 😅
